runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1019,6 +939,434 @@ static iree_status_t iree_hal_hip_device_pepare_async_alloc(
+  callback_data->status = iree_ok_status();
+  *out_data = callback_data;
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t
+iree_hal_hip_device_stream_signal_semaphores_and_add_cleanup(
+    iree_hal_hip_device_t* device, iree_hal_hip_cleanup_thread_t* thread,
+    iree_hal_semaphore_list_t signal_semaphore_list,
+    iree_host_size_t device_ordinal, iree_hal_hip_cleanup_callback_t callback,
+    void* user_data) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_status_t status = iree_ok_status();
+
+  for (iree_host_size_t i = 0; i < signal_semaphore_list.count; ++i) {
+    iree_hal_hip_event_t* event = NULL;
+    iree_status_t status =
+        iree_hal_hip_semaphore_create_event_and_record_if_necessary(
+            signal_semaphore_list.semaphores[i],
+            signal_semaphore_list.payload_values[i],
+            device->topology.devices[device_ordinal].hip_dispatch_stream,
+            device->topology.devices[device_ordinal].device_event_pool, &event);
+
+    if (!iree_status_is_ok(status)) {
+      break;
+    }
+    if (!event) {
+      status =
+          iree_make_status(IREE_STATUS_ABORTED, "the hip event is missing");
+      break;
+    }
+    iree_hal_hip_event_release(event);
+    if (!iree_status_is_ok(status)) {
+      break;
+    }
+  }
+
+  for (iree_host_size_t i = 0;
+       i < signal_semaphore_list.count && iree_status_is_ok(status); ++i) {
+    status = iree_hal_hip_semaphore_notify_forward_progress_to(
+        signal_semaphore_list.semaphores[i],
+        signal_semaphore_list.payload_values[i]);
+  }
+
+  iree_hal_hip_event_t* event = NULL;
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_event_pool_acquire(
+        device->topology.devices[device_ordinal].device_event_pool, 1, &event);
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = IREE_HIP_CALL_TO_STATUS(
+        device->hip_symbols,
+        hipEventRecord(
+            iree_hal_hip_event_handle(event),
+            device->topology.devices[device_ordinal].hip_dispatch_stream));
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_cleanup_thread_add_cleanup(thread, event, callback,
+                                                     user_data);
+  }
+  IREE_TRACE_ZONE_END(z0);
+
+  return status;
+}
+
+typedef struct iree_hal_hip_device_buffer_free_callback_data_t {
+  iree_hal_hip_device_t* device;
+  iree_allocator_t host_allocator;
+  iree_hal_queue_affinity_t queue_affinity;
+  iree_hal_buffer_t* buffer;
+} iree_hal_hip_device_buffer_free_callback_data_t;
+
+static iree_status_t iree_hal_hip_device_make_buffer_free_callback_data(
+    iree_hal_hip_device_t* device, iree_hal_queue_affinity_t queue_affinity,
+    iree_hal_buffer_t* buffer, iree_allocator_t host_allocator,
+    iree_hal_hip_device_buffer_free_callback_data_t** out_data) {
+  *out_data = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_hal_hip_device_buffer_free_callback_data_t* callback_data = NULL;
```

Comment:
```suggestion
  IREE_TRACE_ZONE_BEGIN(z0);
  
  iree_hal_hip_device_buffer_free_callback_data_t* callback_data = NULL;
```

consistently use vertical whitespace - if it is useful to have before then ZONE_END then it should also follow the ZONE_BEGIN